File: test/type_compat/type_compat.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

#include "type_compat.h"

int c_test(void)
{
  int irank;
  int isint, isdbl;
  int idcnt, iicnt;
  int *ibuf;
  double *dbuf;
  double d;
  int i;

  MPI_Initialized(&irank);
  if (!irank) return 1;

  MPI_Comm_rank(MPI_COMM_WORLD, &irank);
  if (irank != 1) return 0;

  MPI_Recv(&isint, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  MPI_Recv(&isdbl, 1, MPI_INT, 0, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  if (isint != sizeof(int)) {
    fprintf(stderr, "// sizeof(int): c: %zu; f: %d;\n",
            sizeof(int), isint);
  }
  if (isdbl != sizeof(double)) {
    fprintf(stderr, "// sizeof(double): c: %zu; f: %d;\n",
            sizeof(double), isdbl);
  }
  MPI_Recv(&iicnt, 1, MPI_INT, 0, 10, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  MPI_Recv(&idcnt, 1, MPI_INT, 0, 11, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

  if (isint > 0 && iicnt > 0) {
    ibuf = (int *)calloc(isint, iicnt);
  } else {
    ibuf = NULL;
  }
  if (isdbl > 0 && idcnt > 0) {
    dbuf = (double *)calloc(isdbl, idcnt);
  } else {
    dbuf = NULL;
  }
  i = 0;
  if (!ibuf || !dbuf) {
    i = 1;
  }
  MPI_Send(&i, 1, MPI_INT, 0, 99, MPI_COMM_WORLD);
  if (i) {
    return -1;
  }

  MPI_Recv(ibuf, iicnt, MPI_INTEGER, 0, 20, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  MPI_Recv(dbuf, idcnt, MPI_DOUBLE_PRECISION, 0, 21, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

  for (i = 0; i < iicnt; ++i) {
    fprintf(stderr, "ibuf[%d] = %5d\n", i, ibuf[i]);
  }

  for (i = 0; i < idcnt; ++i) {
    fprintf(stderr, "dbuf[%d] = %e (%a)\n", i, dbuf[i], dbuf[i]);
  }

  i = 0;
  if (ibuf[0] !=  0) i = 1;
  if (ibuf[1] !=  1) i = 2;
  if (ibuf[2] != -2) i = 3;

  if (dbuf[0] != 1.0) i = 101;
  if (dbuf[1] != 0.5) i = 102;
  if (dbuf[2] != 2.0) i = 103;

  d = fabs(dbuf[3] - 0.1);
  fprintf(stderr, "// fabs(fort(0.1) - c(0.1)) = %e (%a)\n", d, d);
  if (d != 0.0) i = 104;

  return i;
}

File: test/type_compat/type_compat.f90

subroutine test
  use ISO_C_BINDING

  !! Format of module is very different for each compiler
  !! products, but we can make sure the MPI library is compiled with
  !! using compiler.
  use MPI

  implicit none

  double precision :: d(10)
  logical :: linit
  integer :: i(10), ierr, irank
  integer(kind=MPI_ADDRESS_KIND) :: ib, it, isint, isdbl
  integer(c_int) :: ic

  call MPI_Initialized(linit, ierr)
  if (.not. linit) then
     return
  endif

  call MPI_Comm_rank(MPI_COMM_WORLD, irank, ierr)
  if (irank .ne. 0) then
     return
  endif

  call MPI_Get_address(d(0:1), ib, ierr)
  call MPI_Get_address(d(1:2), it, ierr)
  isdbl = it - ib
  call MPI_Get_address(i(0:1), ib, ierr)
  call MPI_Get_address(i(1:2), it, ierr)
  isint = it - ib

  d(1) = 1.0d0           ! 0x1p+0
  d(2) = 0.5d0           ! 0x1p-1
  d(3) = 2.0d0           ! 0x1p+1
  d(4) = 0.1d0           ! 0

  i(1) =  0
  i(2) =  1
  i(3) = -2

  ic = isint
  call MPI_Send(ic, 1, MPI_INT, 1, 0, MPI_COMM_WORLD, ierr)
  ic = isdbl
  call MPI_Send(ic, 1, MPI_INT, 1, 1, MPI_COMM_WORLD, ierr)

  ic = 3
  call MPI_Send(ic, 1, MPI_INT, 1, 10, MPI_COMM_WORLD, ierr)
  ic = 4
  call MPI_Send(ic, 1, MPI_INT, 1, 11, MPI_COMM_WORLD, ierr)
  call MPI_Recv(ic, 1, MPI_INT, 1, 99, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
  if (ic .ne. 0) then
     return
  end if

  call MPI_Send(i, 3, MPI_INTEGER, 1, 20, MPI_COMM_WORLD, ierr)
  call MPI_Send(d, 4, MPI_DOUBLE_PRECISION, 1, 21, MPI_COMM_WORLD, ierr)

  return
end subroutine test

program type_compat
  use ISO_C_BINDING
  use MPI

  implicit none

  interface
     function c_test() bind(c)
       use ISO_C_BINDING
       integer(c_int) :: c_test
     end function c_test
  end interface

  integer :: ierr, irank, nsize
  integer(c_int) :: ci

  ci = 0
  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, irank, ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, nsize, ierr)
  if (irank .eq. 0) then
     call test
  elseif (irank .eq. 1) then
     ci = c_test()
  endif
  call MPI_Finalize(ierr)

  if (ci .ne. 0) then
     print *, 'error ', ci
     stop 1
  end if
  stop
end program type_compat
